### ISSUE TITLE
Update benchmark results for 0.3

### ISSUE THREAD

**Issue by jkeiser** [simdjson/simdjson#591]
We need to rerun the benchmarks to make them real for 0.3:

- [ ] Update competitor versions
- [ ] Add new obvious competitors if any?
- [ ] Rerun the benchmarks
- [ ] Run benchmarks for small files as well as large ones, to make the point that we're fast everywhere
- [ ] Pretty graphs

**Comment by lemire** [comment]
I have moved to the simdjson org two relevant repositories...

https://github.com/simdjson/simdjson_experiments_vldb2019

https://github.com/simdjson/iossimdjson

The first one is a set of docker containers to benchmark simdjson. Using docker is useful because you can ensure that you have always the same system configuration, over different machines. For obvious reasons, it is now somewhat obsolete, but it could be put to good user.

The second one is a small app for ios which one can use for benchmarking purposes.

**Comment by lemire** [comment]
> Update competitor versions

I think that the main competitor remains RapidJSON and it hasn't been updated since 2018. They have little reason to update it considering that it is mostly bug-free at this point, and the performance is, all things said, quite good.

One difference this time around is that since we offer accurate number parsing, we can benchmark against RapidJSON when it uses accurate parsing... But I already included in `parsingcompetition`:

```
$ ./parsingcompetition jsonexamples/numbers.json
getline                                 	:    0.163 cycles per input byte (best)    0.168 cycles per input byte (avg)   19.924 GB/s (error margin: 0.527 GB/s)
simdjson (dynamic mem)                  	:    3.213 cycles per input byte (best)    3.247 cycles per input byte (avg)    1.058 GB/s (error margin: 0.011 GB/s)
simdjson                                	:    3.208 cycles per input byte (best)    3.227 cycles per input byte (avg)    1.060 GB/s (error margin: 0.006 GB/s)
RapidJSON                               	:    7.296 cycles per input byte (best)    7.707 cycles per input byte (avg)    0.467 GB/s (error margin: 0.025 GB/s)
RapidJSON (accurate number parsing)     	:   15.639 cycles per input byte (best)   15.790 cycles per input byte (avg)    0.218 GB/s (error margin: 0.002 GB/s)
RapidJSON (insitu)                      	:    7.205 cycles per input byte (best)    7.320 cycles per input byte (avg)    0.472 GB/s (error margin: 0.007 GB/s)
RapidJSON (insitu, accurate number parsing)	:   15.646 cycles per input byte (best)   15.768 cycles per input byte (avg)    0.218 GB/s (error margin: 0.002 GB/s)
sajson (dynamic mem)                    	:    7.355 cycles per input byte (best)    7.387 cycles per input byte (avg)    0.463 GB/s (error margin: 0.002 GB/s)
sajson                                  	:    6.485 cycles per input byte (best)    6.503 cycles per input byte (avg)    0.525 GB/s (error margin: 0.001 GB/s)
nlohmann-json                           	:   55.619 cycles per input byte (best)   56.017 cycles per input byte (avg)    0.061 GB/s (error margin: 0.000 GB/s)
memcpy                                  	:    0.085 cycles per input byte (best)    0.087 cycles per input byte (avg)   36.279 GB/s (error margin: 0.582 GB/s)
```

**Comment by lemire** [comment]
> Add new obvious competitors if any?

If you build "allparsingcompetition", you have the following...

```
$ ./allparsingcompetition jsonexamples/twitter.json
getline                                 	:    2.448 cycles per input byte (best)    2.464 cycles per input byte (avg)    1.391 GB/s (error margin: 0.009 GB/s)
simdjson (dynamic mem)                  	:    1.396 cycles per input byte (best)    1.442 cycles per input byte (avg)    2.438 GB/s (error margin: 0.077 GB/s)
simdjson                                	:    1.394 cycles per input byte (best)    1.419 cycles per input byte (avg)    2.442 GB/s (error margin: 0.043 GB/s)
RapidJSON                               	:   12.992 cycles per input byte (best)   13.280 cycles per input byte (avg)    0.262 GB/s (error margin: 0.006 GB/s)
RapidJSON (accurate number parsing)     	:   13.498 cycles per input byte (best)   13.533 cycles per input byte (avg)    0.252 GB/s (error margin: 0.001 GB/s)
RapidJSON (insitu)                      	:    8.628 cycles per input byte (best)    8.659 cycles per input byte (avg)    0.395 GB/s (error margin: 0.001 GB/s)
RapidJSON (insitu, accurate number parsing)	:    8.527 cycles per input byte (best)    8.569 cycles per input byte (avg)    0.400 GB/s (error margin: 0.002 GB/s)
sajson (dynamic mem)                    	:    4.286 cycles per input byte (best)    4.338 cycles per input byte (avg)    0.795 GB/s (error margin: 0.010 GB/s)
sajson                                  	:    3.907 cycles per input byte (best)    3.936 cycles per input byte (avg)    0.872 GB/s (error margin: 0.006 GB/s)
nlohmann-json                           	:   36.098 cycles per input byte (best)   36.269 cycles per input byte (avg)    0.094 GB/s (error margin: 0.000 GB/s)
dropbox (json11)                        	:   26.111 cycles per input byte (best)   26.204 cycles per input byte (avg)    0.131 GB/s (error margin: 0.000 GB/s)
fastjson                                	:   11.656 cycles per input byte (best)   11.738 cycles per input byte (avg)    0.292 GB/s (error margin: 0.002 GB/s)
gason                                   	:    4.243 cycles per input byte (best)    4.689 cycles per input byte (avg)    0.803 GB/s (error margin: 0.076 GB/s)
ultrajson                               	:    9.046 cycles per input byte (best)    9.104 cycles per input byte (avg)    0.377 GB/s (error margin: 0.002 GB/s)
jsmn                                    	:   13.286 cycles per input byte (best)   13.334 cycles per input byte (avg)    0.256 GB/s (error margin: 0.001 GB/s)
cJSON                                   	:   11.363 cycles per input byte (best)   11.417 cycles per input byte (avg)    0.300 GB/s (error margin: 0.001 GB/s)
jsoncpp                                 	:   28.130 cycles per input byte (best)   28.461 cycles per input byte (avg)    0.121 GB/s (error margin: 0.001 GB/s)
memcpy                                  	:    0.110 cycles per input byte (best)    0.113 cycles per input byte (avg)   30.361 GB/s (error margin: 0.621 GB/s)
```

[I have reviewed most of these parsers and their limitations/strengths](https://arxiv.org/abs/1902.08318), part of it is in the README.

It is important to understand the compromises that various parsers do.

Then the next question is: do you want to add parsers from other programming languages... C#, Java, and so forth. There are already benchmarking frameworks across languages. Setting that up in a reproducible and fair manner is a bit difficult. We have an issue related to this: on very large files, a parser written in D can be faster, largely because it avoids allocating as many pages. Page allocation becomes the bottleneck when you benchmark a single run of a test that parses a large file when in "small pages" mode. 


I guess that where I am going with this is that you should underestimate the amount of work that could go into this!

**Comment by lemire** [comment]
So if you want to include large files, you have to discuss the cost of page allocation. You can amortize it... Or you can configure your system to use huge pages.

**Comment by jkeiser** [comment]
> Then the next question is: do you want to add parsers from other programming languages... C#, Java, and so forth. 

Not for 0.3 :) We are a C++ framework, we compare against C++.

However, longer term I think our goal is to demonstrate that *up until now, we've all simply been doing parsing wrong.* 2.5x is an absurd multiplier and should not even be possible for something as ubiquitous and simple as JSON. From that point of view, we should at least reference any JSON parser from *any* language that comes anywhere close (if they exist).

> So if you want to include large files, you have to discuss the cost of page allocation. You can amortize it... Or you can configure your system to use huge pages.

Sorry, I think I mean middle-sized then, megabytes not gigabytes :) Basically the range of files that people are likely to encounter. Multi-gigabyte files aren't an official selling point of simdjson yet, so we don't necessarily need to benchmark them.

Basically, at a bare minimum, pick one of the smaller files and benchmark it against the competition, the way I did earlier--or even a range of files; the point is to demonstrate this isn't a one-trick pony. "It probably only handles large files" is going to be a common suspicion given the tagline "Parsing JSON at gigabytes per second." It was for me :)

**Comment by lemire** [comment]
> Sorry, I think I mean middle-sized then, megabytes not gigabytes :) Basically the range of files that people are likely to encounter. Multi-gigabyte files aren't an official selling point of simdjson yet, so we don't necessarily need to benchmark them.

This will be a factor at the megabyte level too. If you have 4kB per page, then 10 MB is still 2500  pages.

See my blog post...

https://lemire.me/blog/2020/01/17/allocating-large-blocks-of-memory-bare-metal-c-speeds/

We are limited at less than 4 GB/s for page allocation... Suppose that you have a parser that runs at 2 GB/s without page allocation, then the final speed with page allocation will be 1.3 GB/s. This is no joke!

**Comment by lemire** [comment]
> Basically, at a bare minimum, pick one of the smaller files and benchmark it against the competition, the way I did earlier--or even a range of files; the point is to demonstrate this isn't a one-trick pony. "It probably only handles large files" is going to be a common suspicion given the tagline "Parsing JSON at gigabytes per second." It was for me :)

I think you had a neat figure in one of the prior issues.

**Comment by jkeiser** [comment]
> We are limited at less than 4 GB/s for page allocation... Suppose that you have a parser that runs at 2 GB/s without page allocation, then the final speed with page allocation will be 1.3 GB/s. This is no joke!

Gotcha. So basically, parsingcompetition is hiding that cost because it picks the best time, and the pages will be fully allocated the second/third time around? Do the other parsers have the ability to keep parser buffers around, or is it just simdjson? If they can keep the parser around, then we should to keep it fair, otherwise this is an advantage of simdjson, and the benchmarks are real. We should put a note next to the results to this effect, probably. I think we already have a note explaining how and when you'd want to do huge page allocation.

> I think you had a neat figure in one of the prior issues.

Yeah, we can definitely reuse that. I also modified parsingcompetition so you can run `parsingcompetition jsonexamples/*.json` and it will run all of the parsers against all the files.

**Comment by lemire** [comment]
> Gotcha. So basically, parsingcompetition is hiding that cost because it picks the best time, and the pages will be fully allocated the second/third time around?


It is mostly hiding the cost, but probably not entirely because you still have page walking and other issues. Still re-using the same parser with allocated memory is the way to go if you can't use huge pages...

Try this...

```
$ git pull origin master
$ ./scripts/ruby/kostya_large.rb
$ ./parse -n 10  scripts/ruby/1.json
$ ./parse -n 10  scripts/ruby/2.json
``` 

It is same data but file 2.json is much smaller. On my test box, I get 1.4 GB/s in the case of the small file and 0.8 GB/s in the case of the large file.

Here is what I get with parsingcompetition... I do not quite achieve 1.4 GB/s, but nearly so... Compare simdjson (dynamic mem) and simdjson. The difference is almost entirely related to page allocation...

```
$ ./parsingcompetition 1.json
getline                                 	:    4.027 cycles per input byte (best)    4.029 cycles per input byte (avg)    0.846 GB/s (error margin: 0.000 GB/s)
simdjson (dynamic mem)                  	:    4.228 cycles per input byte (best)    4.230 cycles per input byte (avg)    0.806 GB/s (error margin: 0.000 GB/s)
simdjson                                	:    2.595 cycles per input byte (best)    2.745 cycles per input byte (avg)    1.313 GB/s (error margin: 0.072 GB/s)
RapidJSON                               	:    8.444 cycles per input byte (best)    8.454 cycles per input byte (avg)    0.404 GB/s (error margin: 0.000 GB/s)
RapidJSON (accurate number parsing)     	:   11.183 cycles per input byte (best)   11.186 cycles per input byte (avg)    0.305 GB/s (error margin: 0.000 GB/s)
RapidJSON (insitu)                      	:    7.366 cycles per input byte (best)    7.368 cycles per input byte (avg)    0.463 GB/s (error margin: 0.000 GB/s)
RapidJSON (insitu, accurate number parsing)	:   10.253 cycles per input byte (best)   10.258 cycles per input byte (avg)    0.332 GB/s (error margin: 0.000 GB/s)
sajson (dynamic mem)                    	:    8.189 cycles per input byte (best)    8.229 cycles per input byte (avg)    0.416 GB/s (error margin: 0.002 GB/s)
sajson                                  	:    5.006 cycles per input byte (best)    5.120 cycles per input byte (avg)    0.681 GB/s (error margin: 0.015 GB/s)
nlohmann-json                           	:   47.564 cycles per input byte (best)   47.606 cycles per input byte (avg)    0.072 GB/s (error margin: 0.000 GB/s)
memcpy                                  	:    0.256 cycles per input byte (best)    0.257 cycles per input byte (avg)   13.320 GB/s (error margin: 0.059 GB/s)
```

**Comment by lemire** [comment]
@jkeiser  It gets tricky, of course... if you have an app and you eat one large JSON document once, then you are going to be limited by page allocation... it is just how it is and a parser that works harder than simdjson to limit memory usage will be better off. If you have a server that routinely eats large JSON documents and can reuse its memory, then you are probably fine.

**Comment by lemire** [comment]
I'd consider closing this issue on the basis of PR https://github.com/simdjson/simdjson/pull/634

**Comment by jkeiser** [comment]
Agreed!
